schematic/schemas/data_model_nodes.py

```diff
@@ -0,0 +1,273 @@
+        )
+        self.properties = self.get_data_model_properties(
+            attr_rel_dict=attribute_relationships_dict
+        )
+        # retrieve a list of relationship types that will produce nodes.
+        self.node_relationships = list(self.edge_relationships_dictionary.values())
+
+    def gather_nodes(self, attr_info: tuple) -> list:
+        """Take in a tuple containing attriute name and relationship dictionary, and find all nodes defined in attribute information.
+        Args:
+            attr_info, tuple: (Display Name, Relationships Dictionary portion of attribute_relationships dictionary)
+        Returns:
+            nodes, list: nodes related to the given node (specified in attr_info).
+        Note:
+            Extracting nodes in this fashion ensures order is preserved.
+        """
+
+        # Extract attribute and relationship dictionary
+        attribute, relationship = attr_info
+        relationships = relationship["Relationships"]
+
+        nodes = []
+        if attribute not in nodes:
+            nodes.append(attribute)
+        for rel in self.node_relationships:
+            if rel in relationships.keys():
+                nodes.extend([node.strip() for node in relationships[rel]])
+        return nodes
+
+    def gather_all_nodes_in_model(self, attr_rel_dict: dict) -> list:
+        """Gather all nodes in the data model, in order.
+        Args:
+            attr_rel_dict, dict: generated in data_model_parser
+                {Attribute Display Name: {
+                        Relationships: {
+                                    CSV Header: Value}}}
+        Returns:
+            all_nodes, list: List of all node display names in the data model preserving order entered.
+        Note:
+            Gathering nodes in this fashion ensures order is preserved.
+        """
+        all_nodes = []
+        for attr_info in attr_rel_dict.items():
+            nodes = self.gather_nodes(attr_info=attr_info)
+            all_nodes.extend(nodes)
+        # Remove any duplicates preserving order
+        all_nodes = list(dict.fromkeys(all_nodes).keys())
+        return all_nodes
+
+    def get_rel_node_dict_info(self, relationship: str) -> Optional[tuple[str, dict]]:
+        """For each display name get defaults for nodes.
+        Args:
+            relationship, str: relationship key to match.
+        Returns:
+            rel_key, str: relationship node label
+            rel_node_dict, dict: node_attr_dict, from relationships dictionary for a given relationship
+        TODO: Move to data_model_relationships.
+        """
+        for k, v in self.data_model_relationships.relationships_dictionary.items():
+            if k == relationship:
+                if "node_attr_dict" in v.keys():
+                    rel_key = v["node_label"]
+                    rel_node_dict = v["node_attr_dict"]
+                    return rel_key, rel_node_dict
+
```

Comment:
I'm a little confused about this method. Is 
```
for k, v in self.data_model_relationships.relationships_dictionary.items():
            if k == relationship:
```
the same as `self.data_model_relationships.relationships_dictionary[relationship]`, I'm wondering why you need to iterate through the whole dict instead of treating using the key to get the value out of the dict as normal?

Reply:
I'd write this as 

```
value =  self.data_model_relationships.relationships_dictionary.get(relationship)
if value:
    value2 = value.get("node_attr_dict")
    if value2:
         rel_key = v["node_label"]
         rel_node_dict = v["node_attr_dict"]
         return rel_key, rel_node_dict
```

Unless I'm misunderstanding what the method is doing

Reply:
I think that could probably work. I tend to get a little nervous about key errors so add additional checks but in this case for making a node_dict I think we can assume those keys are present.

Reply:
Yep, that's why I'm using `dict.get(relationship)`. This safely gets the value in the dict associated with the key. If they key exists it returns the value, if it doesn't then it returns None. Therefore there's no Key error ever.

https://www.w3schools.com/python/ref_dictionary_get.asp